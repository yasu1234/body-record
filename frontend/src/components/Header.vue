<script setup>
import { ref } from 'vue';
import { useRouter } from 'vue-router'
import Cookies from 'js-cookie';

const router = useRouter();

const props = defineProps({
    isLogin: Boolean
})

const menuList = ref([
  { id: 1, label: "ユーザー設定", isLogin: true },
  { id: 2, label: "設定", isLogin: false },
  { id: 3, label: "ログアウト", isLogin: true }
]);

const showDropdown = ref(false);

function showAccountIntroduction() {
    router.push({ name: 'AccountInteroduction'})
}

function toggleDropdown() {
  showDropdown.value = !showDropdown.value;
}

function showMenu(menu) {
}
</script>

<template>
    <header>
        <h2 class="title"><router-link to="/" class="title">体レコ.com</router-link></h2>
        <nav>
            <div class="navigation-buttons">
                <button v-if="!isLogin" class="account-introducton-button" @click="showAccountIntroduction">会員登録・ログイン</button>
                <div @click="toggleDropdown">
                    <button class="user-button"><img src="../assets/image/user-placeholder.png" alt="ユーザー" class="user-image"></button>
                </div>                
            </div>
            <div v-if="showDropdown" class="dropdown-menu">
                <div v-for="menu in menuList" @click="showMenu(menu)">
                    {{ menu.label }}
                </div>
            </div>
        </nav>
    </header>
</template>

<style>
header {
    width: 100%;
    height: 80px;
    background: #f1f3ef;
    padding-left: 30px;
    padding-right: 30px;
    box-sizing: border-box;
    display: flex;
    align-items: center;
    justify-content: space-between;
    position: fixed;
    top: 0;
}
.title {
    color:#ffa600;
}
.navigation-buttons {
    display: flex;
    justify-content: space-between;
}
.account-introducton-button {
    background: #ffa500;
    margin-right: 30px;
}
.user-button {
    padding: 0;
    background: transparent;
    border: 1px solid #CCC;
    border-radius: 50%;
}
.user-image {
    width: 40px;
    height: 40px;
}

.user-dropdown {
  position: relative;
  display: inline-block;
}

.dropdown-menu {
  position: absolute;
  background-color: #f1f1f1;
  min-width: 160px;
  box-shadow: 0px 8px 16px 0px rgba(0,0,0,0.2);
  z-index: 1;
  right: 5px;
  top: 80px;
  display: flex;
  flex-direction: column;
  border-radius: 4px;
}
.menu-item {
  padding: 6px;
}
</style>