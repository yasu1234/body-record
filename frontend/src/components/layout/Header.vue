<script setup>
import { ref, onMounted } from 'vue';
import { useRouter, useRoute } from 'vue-router'
import Cookies from 'js-cookie';
import axios from 'axios';

const router = useRouter();

const isLogin = ref(false);
const userId = ref(0);

onMounted(() => {
    checkLogin();
});

const menuList = ref([
  { id: 1, label: "マイページ", isLogin: true },
  { id: 2, label: "設定", isLogin: true },
  { id: 3, label: "ログアウト", isLogin: true },
  { id: 4, label: "問い合わせ", isLogin: false }
]);

const checkLogin = async () => {
  try {
    const res = await axios.get(import.meta.env.VITE_APP_API_BASE + '/api/v1/users/check_login', {
      headers: {
        'access-token' : Cookies.get('accessToken'),
        'client':Cookies.get('client'),
        'uid': Cookies.get('uid'),
      },
    })

    isLogin.value = (res.data.user !== null)
    userId.value = res.data.user.id
  } catch (error) {
    isLogin.value = false
  }
}

const showDropdown = ref(false);

function showMyPage() {
    router.push({ name: 'UserProfile', params: { id: userId.value }})
}

function showAccountIntroduction() {
    router.push({ name: 'AccountInteroduction'})
}

function showSetting() {
    router.push({ name: 'Setting'})
}

function showContact() {
    router.push({ name: 'Contact'})
}

function toggleDropdown() {
  showDropdown.value = !showDropdown.value;
}

function showMenu(menu) {
    switch(menu.id) {
        case 1:
            showMyPage();
            break
        case 2:
            showSetting();
            break
        case 3:
            logout();
            break
        case 4:
            showContact();
            break
        default:
            break
    }
}

const logout = async () => {
    try {
        const res = await axios.delete(import.meta.env.VITE_APP_API_BASE + '/api/v1/auth/sign_out', {
            headers: {
                'access-token' : Cookies.get('accessToken'),
                'client':Cookies.get('client'),
                'uid': Cookies.get('uid')
            }
        })
        Cookies.remove('accessToken')
        Cookies.remove('client')
        Cookies.remove('uid')

        showHomeThenRelaod()
    } catch (error) {
        console.log({ error })
    }
}

const showHomeThenRelaod = async () => {
  await router.push({ name: 'Home'})
  router.go(0)
}
</script>

<template>
    <header>
        <h2 class="title"><router-link to="/" class="title">In-Body.com</router-link></h2>
        <nav>
            <div class="navigation-buttons">
                <button v-if="!isLogin" class="account-introducton-button" @click="showAccountIntroduction">会員登録・ログイン</button>
                <div @click="toggleDropdown">
                    <button class="user-button"><img src="../../assets/image/user-placeholder.png" alt="ユーザー" class="user-image"></button>
                </div>                
            </div>
            <div v-if="showDropdown" class="dropdown-menu">
                <button v-for="menu in menuList" class="menu-item" @click="showMenu(menu)">
                    {{ menu.label }}
                </button>
            </div>
        </nav>
    </header>
</template>

<style>
header {
    width: 100%;
    height: 80px;
    background: #f1f3ef;
    padding-left: 30px;
    padding-right: 30px;
    box-sizing: border-box;
    display: flex;
    align-items: center;
    justify-content: space-between;
}
.title {
    color:#ffa600;
}
.navigation-buttons {
    display: flex;
    justify-content: space-between;
}
.account-introducton-button {
    background: #ffa500;
    margin-right: 30px;
}
.user-button {
    padding: 0;
    background: transparent;
    border: 1px solid #CCC;
    border-radius: 50%;
}
.user-image {
    width: 40px;
    height: 40px;
}

.user-dropdown {
  position: relative;
  display: inline-block;
}

.dropdown-menu {
  position: absolute;
  background-color: #f1f1f1;
  min-width: 160px;
  box-shadow: 0px 8px 16px 0px rgba(0,0,0,0.2);
  z-index: 1;
  right: 5px;
  top: 80px;
  display: flex;
  flex-direction: column;
  border-radius: 4px;
}
.menu-item {
  padding: 6px;
}
</style>