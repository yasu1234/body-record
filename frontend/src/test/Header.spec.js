import { shallowMount } from '@vue/test-utils'
import Header from '@/components/layout/Header.vue'
import Cookies from 'js-cookie'

jest.mock("axios")
jest.mock('js-cookie')
jest.mock('vue-router', () => {
  const router = {
    push: jest.fn(),
    pathname: '/',
    route: '',
    asPath: '',
    currentRoute: {
      value: {
        fullPath: '/',
      }
    },
    query: {},
    events: {
      on: jest.fn(),
      off: jest.fn(),
      emit: jest.fn(),
    },
  };
  return {
    useRouter: jest.fn().mockReturnValue(router),
  };
});

jest.mock('@/js/axios.js', () => ({
  axiosInstance: {
    get: jest.fn(),
    delete: jest.fn(),
    interceptors: {
      request: { use: jest.fn() },
      response: { use: jest.fn() },
    },
  },
  setupInterceptors: jest.fn(),
}));

describe('Header.vue', () => {
  let wrapper;
  let mockedAxiosInstance;

  beforeEach(() => {
    wrapper = shallowMount(Header, {
      global: {
        stubs: ["router-link", "router-view"]
      }
    })
  })

  it('SnapShotテスト', () => {
    expect(wrapper.element).toMatchSnapshot();
  })

  it('未ログインの場合には会員登録・ログインボタンが表示される', async () => {
    jest.mock("@/js/axios.js", () => ({
      axiosInstance: {
        get: jest.fn(),
        post: jest.fn(),
        delete: jest.fn(),
        interceptors: {
          request: { use: jest.fn() },
          response: { use: jest.fn() },
        },
      },
      setupInterceptors: jest.fn(),
    }));

    // モックされたaxiosInstanceを取得
    const axiosModule = require("@/js/axios.js");
    mockedAxiosInstance = axiosModule.axiosInstance;
    mockedAxiosInstance.get.mockResolvedValueOnce({ data: { user: null } })

    expect(wrapper.find('.account-introducton-button').exists()).toBe(true)
  })

  it('ログイン済みの場合には会員登録・ログインボタンが表示されない', async () => {
    const user = { id: 1 }

    jest.mock("@/js/axios.js", () => ({
      axiosInstance: {
        get: jest.fn(),
        post: jest.fn(),
        delete: jest.fn(),
        interceptors: {
          request: { use: jest.fn() },
          response: { use: jest.fn() },
        },
      },
      setupInterceptors: jest.fn(),
    }));

    const axiosModule = require("@/js/axios.js");
    mockedAxiosInstance = axiosModule.axiosInstance;
    mockedAxiosInstance.get.mockResolvedValue({ data: { user: user } })

    expect(wrapper.find('.user-button').exists()).toBe(true)
  })

  it('ログイン済みの場合にはドロップダウンメニューが4つ表示される', async () => {
    const user = { id: 1 }

    jest.mock("@/js/axios.js", () => ({
      axiosInstance: {
        get: jest.fn(),
        post: jest.fn(),
        delete: jest.fn(),
        interceptors: {
          request: { use: jest.fn() },
          response: { use: jest.fn() },
        },
      },
      setupInterceptors: jest.fn(),
    }));

    const axiosModule = require("@/js/axios.js");
    mockedAxiosInstance = axiosModule.axiosInstance;

    mockedAxiosInstance.get.mockResolvedValue({ data: { user: user } })

    await wrapper.vm.checkLogin()
    await wrapper.find('.user-button').trigger('click')

    const menuItems = wrapper.findAll('.menu-item')
    expect(menuItems.length).toBe(4)
    expect(wrapper.find('.dropdown-menu').exists()).toBe(true)
  })

  it('ログアウトするとCookieのトークン情報がなくなる', async () => {
    const user = { id: 1 }

    jest.mock("@/js/axios.js", () => ({
      axiosInstance: {
        get: jest.fn(),
        post: jest.fn(),
        delete: jest.fn(),
        interceptors: {
          request: { use: jest.fn() },
          response: { use: jest.fn() },
        },
      },
      setupInterceptors: jest.fn(),
    }));

    const axiosModule = require("@/js/axios.js");
    mockedAxiosInstance = axiosModule.axiosInstance;

    mockedAxiosInstance.get.mockResolvedValueOnce({ data: { user: user } })
    
    await wrapper.vm.checkLogin()
    await wrapper.find('.user-button').trigger('click')
  
    const menuItems = wrapper.findAll('.menu-item')
    expect(menuItems.length).toBeGreaterThan(3)
  
    await menuItems.at(3).trigger('click')

    mockedAxiosInstance.delete.mockResolvedValueOnce({})
  
    expect(Cookies.get('accessToken')).toBeUndefined()
    expect(Cookies.get('client')).toBeUndefined()
    expect(Cookies.get('uid')).toBeUndefined()
  })
})
