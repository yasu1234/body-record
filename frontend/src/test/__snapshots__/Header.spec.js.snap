// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header.vue SnapShotテスト 1`] = `
<header>
  <h2
    class="title"
  >
    <router-link-stub
      class="title"
      to="/"
    />
  </h2>
  <nav>
    <div
      class="navigation-menus"
    >
      <button
        class="account-introducton-button mr-7"
        style=""
      >
         会員登録・ログイン 
      </button>
      <div>
        <button-stub
          aria-label="ユーザー"
          class="user-button"
          icon="pi pi-user"
          pt="[object Object]"
          rounded=""
          severity="info"
        />
      </div>
    </div>
    <!--v-if-->
  </nav>
</header>
`;
